### DISCUSSION TITLE
Detecting bus routes from a trajectory

### DISCUSSION THREAD

**Discussion by dabreegster** [anitagraser/movingpandas#229] [Q&A]
Hi! I'm currently working on [an open source tool](https://github.com/dabreegster/bus_spotting) that analyzes data from a bus operator. I hit an interesting challenge working with trajectory data, and I thought somebody in this community might have ideas how to solve it (or maybe even knows of an existing solution -- I doubt this is a new problem!)

I'm not using MovingPandas in my project (because I'm not working in Python), but the library has been a huge inspiration/reference. If anyone can help solve the problem below in terms of MovingPandas, that'd be plenty for me to figure out how to translate things over.

# The problem

I have two datasets, with a sample at https://gist.github.com/dabreegster/e5399e398f817f3ca2ae10cb98fc6f0f

1) Raw trajectories describing the motion of individual buses over the course of a day. I've included one as a CSV (with x, y, time) in the gist.
2) A set of possible bus routes that the vehicle might serve over the day. Each one has a line-string shape, and an ordered sequence of stops that the bus is supposed to visit in order. I've saved this as GeoJSON in the gist.

I want to combine the two and figure out what routes a bus actually served through the day. A bus serves multiple routes over a day, often two routes that're opposites of each other (one direction, then another). It might also do various off-route things like visit a repair station, and in practice, I've seen examples where a bus apparently aborts a route early, or skips a few stops in the middle.

The output I'm trying to get would be something like:

- the bus serves route "123" from 7am to 9am
  - stop 1 at 7am, stop 2 at 7:05, stop 3 at 7:06....
- route "456" from 9am to 10
  - stop 1 at 9am, stop 2 at 9:10...
- route "123" from 10:30 to 12
  - note there's a 30 minute gap before serving this route

# My solution so far

I tried [a bunch of ideas](https://github.com/dabreegster/bus_spotting/issues/9) that haven't really panned out. The one that appears to work the best:

**subroutine**: given a vehicle and a possible route, return all actual trips it takes.  (actual trip = "stop 1 at t1, stop 2 at t2, ..."
- Find all times the trajectory is within **30 meters** of each stop
- Try to stitch those together into sequences of stops in order. Start with the earliest time at stop 1. Find the earliest time at stop 2 that's > time1. Then the earliest time at stop 3 that's > time2. etc.

So in other words, the intermediate state is something like:

stop1: 7:00, 7:01, 9:00, 9:05
stop2: 6:50, 7:01, 8:30
stop3: ...

There's noise here, because the trajectory might pass close to some stops without actually stopping there. I've had to bump up the distance threshold to 30 meters based on some examples I've found where a bus almost definitely made a certain stop (verified by watching an animation), but due to GPS inaccuracies and stuff, the distance is a bit higher.

**subroutine**: given a vehicle, return all trips it takes over the whole day, for any possible route.
- Use some other data to figure out routes that possibly match (usually about 4 or 5 per vehicle)
- Call the above
- A vehicle can't be doing two trips at the same time. So walk through the total list of trips and pick some that don't overlap. I've been sorting by the full trip duration and preferring to slot in shorter trips first, because the first subroutine often returns some incorrect-looking trips that last 3 hours

The bug I see most often is a single trip lasting an absurd duration -- like 3 hours, when it normally takes 1 hour. When I look at the delay between each stop, most will be a few minutes, and one will suddenly spike to a few hours. What happens here is often that the first subroutine misses one of the times that a bus is close to a stop, and so the matching greedily picks up the **next** time that bus serves the same route.

Overall, this approach works really well sometimes, but it's brittle / dependent on that 30 meter distance parameter. And the first subroutine also gets confused if the earliest time at the very first stop is a false match.

**Comment by anitagraser** [comment]
Very interesting problem, Dustin. I wonder if the buses always do the whole route? In that case, we might just check for the first & last stop and then compare the trajectory between these locations to the planned route geometry.

**Reply by dabreegster** [reply]
I've seen a few cases in my sample dataset with buses ending a route early, but mostly they follow the whole route. Another common pattern is stopping at the 2nd- or 3rd-to-last stop and using it as a parking zone for a while, before beginning the next route. In that case, it looks like the delay between stop `n-1` and `n` is really high.

IIUC your proposal is to:

1) find all times where the trajectory is within X meters of the first and last stop
2) slice the trajectory between all of those start and end times
3) use some kind of distance function between the two line-strings to see if there's a good match

And once there's a good match, figure out when the bus is within X meters of each stop along the way, to fill out the full schedule. Between each clipped trajectory now, there should only be one visit to each stop in the route.

Thanks for the ideas! I hope to have the project a bit more cleaned up in a few weeks; I'll link to any final documentation when it's ready.
